File: app/schemas/inventory.py
"""
Pydantic schemas for inventory management.
"""

from pydantic import BaseModel, Field
from datetime import date
from typing import Optional, List
from decimal import Decimal


class InventoryBase(BaseModel):
    """Base inventory schema."""
    room_type_id: int
    date: date
    available_rooms: int = Field(..., ge=0)
    price: Decimal = Field(..., gt=0)


class InventoryRead(InventoryBase):
    """Schema for reading inventory data."""
    id: int
    room_type_name: Optional[str] = None
    
    class Config:
        from_attributes = True


class InventoryQuery(BaseModel):
    """Schema for querying inventory by date range."""
    start_date: date
    end_date: date
    room_type_id: Optional[int] = None


class InventoryUpdate(BaseModel):
    """Schema for updating inventory."""
    available_rooms: Optional[int] = Field(None, ge=0)
    price: Optional[Decimal] = Field(None, gt=0)


class InventoryAvailability(BaseModel):
    """Schema for availability check response."""
    room_type_id: int
    room_type_name: str
    date: date
    available_rooms: int
    price: Decimal


class DateRangeAvailability(BaseModel):
    """Availability summary for a date range."""
    room_type_id: int
    room_type_name: str
    start_date: date
    end_date: date
    min_available: int
    total_price: Decimal
    daily_breakdown: List[InventoryAvailability]

File: frontend/src/services/inventory.ts
import { api } from '../lib/axios';

export interface DateRangeAvailability {
    room_type_id: number;
    room_type_name: string;
    start_date: string;
    end_date: string;
    min_available: number;
    total_price: number;
    daily_breakdown: any[]; // define stricter if needed
}

export interface InventoryRead {
    id: number;
    room_type_id: number;
    room_type_name: string;
    date: string;
    available_rooms: number;
    price: number;
}

export interface InventoryUpdate {
    available_rooms?: number;
    price?: number;
}

export const inventoryService = {
    getAvailability: async (start: string, end: string, roomTypeId?: number) => {
        const params = new URLSearchParams({ start, end });
        if (roomTypeId) params.append('room_type_id', roomTypeId.toString());

        const response = await api.get<DateRangeAvailability[]>(`/inventory?${params.toString()}`);
        return response.data;
    },

    getDetailed: async (start: string, end: string, roomTypeId?: number) => {
        const params = new URLSearchParams({ start, end });
        if (roomTypeId) params.append('room_type_id', roomTypeId.toString());

        const response = await api.get<InventoryRead[]>(`/inventory/detailed?${params.toString()}`);
        return response.data;
    },

    update: async (id: number, data: InventoryUpdate) => {
        const response = await api.put<InventoryRead>(`/inventory/${id}`, data);
        return response.data;
    },

    generate: async (roomTypeId: number, daysAhead: number = 90) => {
        const response = await api.post(`/inventory/generate/${roomTypeId}?days_ahead=${daysAhead}`);
        return response.data;
    }
};
